### PR TITLE
fix: handle undefined values of `allowSignup` when they are not set in the login url or app options

### LINKED ISSUE #369: [BUG]: Using `oauth.allowSignup: false` has no effect while using `createNodeMiddleware`
### What happened?

When I am using `createNodeMiddleware` I noticed that the authorization URL is using `allowSignup` as `true` by default. In order to disable that I added `oauth.allowSignup: false` in the App options as such
```js
function setupGitHub() {
  return new App({
    appId: process.env.GITHUB_APP_ID!,
    privateKey: readFileSync(process.env.GITHUB_KEY!, 'utf8'),
    webhooks: {
      secret: process.env.GITHUB_WEBHOOK_SECRET!
    },
    oauth: {
      clientId: process.env.GITHUB_CLIENT_ID!,
      clientSecret: process.env.GITHUB_CLIENT_SECRET!,
      allowSignup: false,
      redirectUrl: process.env.GITHUB_REDIRECT_URL
    }
  })
}
```
But this had no effect.

This seems to me because I didnt add the `allowsignup` parameter to the `login` request. The relevant line where the bug is happening is https://github.com/octokit/oauth-app.js/blob/728b032e3c6233350bb76744e5940c28d9cba01e/src/methods/get-web-flow-authorization-url.ts#L24
and the `options.allowSignup` in the above line is true because
https://github.com/octokit/oauth-app.js/blob/728b032e3c6233350bb76744e5940c28d9cba01e/src/middleware/handle-request.ts#L79
and `query.allowSignup !== "false"` evaulates to `true` if the login request doesn't contain the `allowSignup` parameter.

### Versions

Octokit.js@2.0.10 and @octokit/oauth-app@4.1.0

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by SayakMukhopadhyay**
Resolves #369

----

### Before the change?

* Setting the app option of `allowSignup` would still create a URL with `allow_signup=true`
* The only way to set that option was to call the login URL with `allowSignup=false`
* This means the app option was not having any effect

### After the change?

* If the login URL or the app option, doesn't set the `allowSignup` setting, its defaulted to `true` as previous
* If the login URL sets the `allowSignup` but the app options doesn't, the value is taken from the login URL, whether true or false
* If the login URL doesn't set the `allowSignup` but the app options does, the value is taken from the app options, whether true or false
* If both the login URL and the app options sets the value of  `allowSignup`, the final value is an `OR` operation between the 2 values

### Other information

* None

----

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:  

* N/A

### Pull request type

 
 

Please add the corresponding label for change this PR introduces:
- Bugfix: `Type: Bug`

----

### CODE REVIEW

**Comment by SayakMukhopadhyay** [comment]
Do note, that I am unable to add a label and hence the failing check.

**Review by gr2m — APPROVED** [review]
Another great PR, thank you!

**Comment by baoshan** [comment]
I’m not sure if @gr2m or other maintainers saw my comments. So I’m copying it again:

These lines are hard to reasoning and maintain.:

```
 let allowSignup: boolean;
 if (options.allowSignup === undefined && state.allowSignup === undefined) {
   allowSignup = true;
 } else if (
   options.allowSignup === undefined &&
   state.allowSignup !== undefined
 ) {
   allowSignup = state.allowSignup;
 } else if (
   state.allowSignup === undefined &&
   options.allowSignup !== undefined
 ) {
   allowSignup = options.allowSignup;
 } else {
   allowSignup = options.allowSignup || state.allowSignup;
 }
```

They are literally the same as:

`state.allowSignup ?? options.allowSignup ?? true`

Second, @SayakMukhopadhyay says:

> If the login URL sets the allowSignup but the app options doesn't, the value is taken from the login URL, whether true or false

But the implementation and test case is the opposite:

```
it("GET /api/github/oauth/login?allowSignup=false with allowSignup set to true", async () => {
...
expect(url.searchParams.get("allow_signup")).toEqual("true")
}
```

I think `options.allowSignup ?? state.allowSignup ?? true` is the correct implementation here.

**Comment by wolfy1339** [comment]
> I’m not sure if @gr2m or other maintainers saw my comments. So I’m copying it again:

I didn't see any comments at all, anywhere. 🤔

> These lines are hard to reason about and maintain.:
> [ ... ]
> 
> They are literally the same as:
> 
> `state.allowSignup ?? options.allowSignup ?? true`

I agree those lines are hard to follow, and should be simplified.

**Comment by SayakMukhopadhyay** [comment]
> Second, @SayakMukhopadhyay says:
> 
> > If both the login URL and the app options sets the value of allowSignup, the final value is an OR operation between the 2 values
> 
> I’m not sure where this is requirement comes from. Could you provide a use case?

That check was how things were before my implementation and hence I did it as such to maintain backwards compatibility. But I agree with the sentiments that the huge block of conditional code is a bit janky but I believe `state.allowSignup ?? options.allowSignup ?? true` won't return true when `state.allowSignup` is false and `options.allowSignup` is true.

Still, if you folks are ok with the nullish coalescnce operator, I will create a quick PR to change it.

**Comment by baoshan** [comment]
Sorry for forgetting to submit my review yesterday.

For other options such as `scopes` and (newly introduced) `redirectUrl`, the initial constructor configuration (e.g., `state`) takes precedence over `options`.

I believe for `scopes` and `redirectUrl ` this is the **only secure way**.

If there’s no real-world use case, I prefer we keep `allowSignup` in align with other options, which means `state.allowSignup ?? options.allowSignup ?? true` returns `false` when `state.allowSignup` is `false`.

**Comment by gr2m** [comment]
> I agree those lines are hard to follow, and should be simplified

@baoshan Yes I agree. But I also prefer to merge a working pull request and iterate on it. The tests are there already so should be a fun exercise. Would you like to send a follow up pull request and we continue the discussion there?

**Comment by SayakMukhopadhyay** [comment]
I have created a draft PR #373 to discuss this.
